cmd/dsl/main.go

```diff
@@ -40,37 +44,56 @@ func (t TestError) Error() string {
 }
 
 type Simulator struct {
-	puppetMap    map[string]Puppet
-	portCounter  int
-	instr        Instruction
-	instructions []Instruction
-	basePort     int
+	puppetMap       map[string]Puppet
+	puppetDir       string
+	portCounter     int
+	instr           Instruction
+	instructions    []Instruction
+	basePort        int
+	implementations map[string]string
 }
 
-const (
-	PUPPETSCRIPT = "/home/cblgh/code/netsim-experiments/ssb-server/start-nodejs-puppet.sh"
-)
-
-func startPuppet(p Puppet) error {
-	filename := fmt.Sprintf("./log-%s.txt", p.name)
-	outfile, err := os.Create(filename)
+func startPuppet(s Simulator, p Puppet, shim string) error {
+	filename := path.Join(s.puppetDir, fmt.Sprintf("%s.txt", p.name))
+	logfile, err := os.Create(filename)
 	if err != nil {
 		return TestError{err: err, message: "could not create log file"}
 	}
-	defer outfile.Close()
-	cmd := exec.Command(PUPPETSCRIPT, strconv.Itoa(p.instanceID))
-	cmd.Stderr = outfile
-	cmd.Stdout = outfile
+	defer logfile.Close()
+	var cmd *exec.Cmd
+	// currently the simulator has a requirement that each language implementation folder must contain a sim-shim.sh file
+	// sim-shim.sh contains logic for starting the corresponding sbot correctly.
+	// e.g. reading the passed in ssb directory ($1) and port ($2)
+	cmd = exec.Command(path.Join(s.implementations[shim], "sim-shim.sh"), p.directory, strconv.Itoa(p.Port))
+	cmd.Stderr = logfile
```

Comment:
just a nudge to play with here. If you do `io.MultiWriter(os.Stderr, logfile)` you will see all the output as it happens.


Might want to control this with a `-verbose` flag tho.